core/src/chain_prover.rs

```diff
@@ -0,0 +1,415 @@
+use circuits::header_chain::{HeaderChainCircuitInput, HeaderChainPrevProofType};
+use risc0_zkvm::{AssumptionReceipt, ExecutorEnv, Receipt};
+
+// Checks this amount of previous blocks if not synced with blockchain.
+// TODO: Get this from config file.
+const DEEPNESS: u64 = 5;
+
+/// Possible fetch results.
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum BlockFetchStatus {
+    /// In sync with blockchain.
+    UpToDate,
+    /// Current tip is fallen behind with `height` and `hash`.
+    FallenBehind(u64, BlockHash),
+    /// Current saved tip (with `height` specified) is too far behind the actual tip.
+    OutOfBounds(u64),
+    /// Current tip is considered a fork with `height` and `hash`.
+    Fork(u64, BlockHash),
+}
+
+#[derive(Debug, Clone)]
+pub struct ChainProver<R>
+where
+    R: RpcApiWrapper,
+{
+    rpc: ExtendedRpc<R>,
+    db: Database,
+}
+
+impl<R> ChainProver<R>
+where
+    R: RpcApiWrapper,
+{
+    pub async fn new(config: &BridgeConfig, rpc: ExtendedRpc<R>) -> Result<Self, BridgeError> {
+        let db = Database::new(config).await?;
+
+        Ok(ChainProver { rpc, db })
+    }
+
+    /// Get the proof of a block.
+    ///
+    /// # Parameters
+    ///
+    /// - `height`: Target block height
+    /// - `hash`: [Optional] Target block hash
+    pub async fn get_header_chain_proof(_height: u64, _hash: Option<block::BlockHash>) {
+        todo!()
+    }
+
+    /// Starts a background task that syncs current database to active
+    /// blockchain and does proving.
+    pub fn start_block_prover(&'static self) {
+        tokio::spawn(async move {
+            loop {
+                match self.check_for_new_blocks().await.unwrap() {
+                    BlockFetchStatus::UpToDate => (),
+                    BlockFetchStatus::FallenBehind(_block_height, _block_hash) => {
+                        self.sync_blockchain().await.unwrap()
+                    }
+                    _ => panic!("Hapi yuttun"),
+                };
+
+                let proof_options = HeaderChainCircuitInput {
+                    method_id: [0; 8],
+                    prev_proof: HeaderChainPrevProofType::GenesisBlock,
+                    block_headers: vec![],
+                };
+                let proof = self
+                    .prove_block(proof_options, None::<Receipt>)
+                    .await
+                    .unwrap();
+
+                // TODO: Use real block hash.
+                self.db
+                    .save_block_proof(None, BlockHash::all_zeros(), proof)
+                    .await
+                    .unwrap();
+            }
+        });
+
+        todo!()
+    }
+
+    /// Checks current status of the database against latest active blockchain
+    /// tip.
+    ///
+    /// # Returns
+    ///
+    /// - [`BlockFetchStatus`]: Status of the current database tip
+    async fn check_for_new_blocks(&self) -> Result<BlockFetchStatus, BridgeError> {
```

Comment:
A better way to do is i think, having a function get_blocks_to_prove
Which will return an error if it is to deep, in success, it can return the blockheaders and chain_state of the last proven block

Reply:
I think this is has a conflict with the 2 thread approach, mentioned below. Other thread probably won't have this return value in any point of time.

Reply:
Yes, we can delete this